### PR TITLE
Pread/Pwrite redefinitions

### LINKED ISSUE #78: Writing into a dynamic container
First of all, I just wanted to say that I really like `scroll`! The context system and type-based parsing and everything works perfectly for my use case.

However, there is one thing that I can't seem to figure out how to do - writing types into a container like a `Vec` where the full size of the type is unknown.

Consider this (admittedly contrived) example:
```rust
pub struct MyType;
impl TryIntoCtx<'_> for MyType {
    type Error = scroll::Error;

    fn try_into_ctx(self, into: &mut [u8], _: ()) -> Result<usize, Self::Error> {
        let offset = &mut 0;

        if /* something that isn't statically determined */
        {
            into.gwrite_with(1u16, scroll::LE)?;
        } else {
            into.gwrite_with(1u32, scroll::LE)?;
        }
        
        Ok(*offset)
    }
}
```

If I pre-allocated a `let mut buf = [0u8; 4]` to cover the u32 case, then I'd be wasting space on the u16 case.

What I would love to do is something like this:
```rust
let mut buf = vec![];
buf.gwrite(MyType, 0)?;
```
where the `Vec<u8>` only allocates as-needed.

Is this something that's possible, potentially with a new container trait?

### PR DESCRIPTION
**Pull request by nickbclifford**
This PR adds a new type, `DynamicBuffer`, that automatically expands when `Pwrite`ing types into it. Details in the doc comments.

It also includes a relaxed definition of the `Pwrite` trait that, while technically an API break, is source-compatible and makes implementing custom containers easier.

Resolves #78.

### CODE REVIEW

**Comment by nickbclifford** [comment]
I was considering adding a `TryIntoCtx<_, DynamicBuffer>` impl to the `Pwrite` derive macro as well, but (no offense intended!) the proc macro doesn't seem to be documented well and I didn't really fully understand the specifics of how it worked, so I ultimately decided not to.

**Comment by nickbclifford** [comment]
Could I get a review @m4b?

File: src/buffer.rs [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+        self.buffer.index(self.start_offset + index)
+    }
+}
+impl IndexMut<usize> for DynamicBuffer {
+    fn index_mut(&mut self, mut index: usize) -> &mut Self::Output {
+        index += self.start_offset;
+
+        if let Some(diff) = index.checked_sub(self.buffer.len()) {
+            self.buffer
+                .extend(core::iter::repeat(0).take(core::cmp::max(self.alloc_increment, diff + 1)));
```
Comment (m4b) [inline src/buffer.rs:109]: this is a bit hard to read; needs a comment at least

File: src/buffer.rs [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        }
+
+        self.write_end = self.write_end.max(index);
```
Comment (m4b) [inline src/buffer.rs:112]: shouldn't this only be updated if the buffer gets resized? if yes, I'd suggest to put it into that block

File: src/buffer.rs [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    }
+
+    #[test]
+    fn int_write() {
+        let mut buf = DynamicBuffer::new();
+
+        buf.pwrite_with(0x1234u16, 0, Endian::Little).unwrap();
+        buf.pwrite_with(0x5678i16, 2, Endian::Big).unwrap();
+
+        assert_eq!(buf.get(), [0x34, 0x12, 0x56, 0x78]);
```
Comment (m4b) [inline src/buffer.rs:209]: i haven't thought deeply, but be careful to think about how this test would pass on big endian machine. In other places in scroll there sometimes need to be `cfg(target_endian = "little")` etc.; I think you're ok here but something to think about for a minute, because in my experience easy to make subtle mistake with endianness in tests and it fails years later when someone tries to build it on some big endian machine :)

File: src/buffer.rs [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+impl<Ctx: Copy, E> Pwrite<Ctx, E> for DynamicBuffer {
+    type Buffer = Self;
+
+    fn pwrite_with<N: TryIntoCtx<Ctx, Self::Buffer, Error = E>>(
+        &mut self,
+        n: N,
+        offset: usize,
+        ctx: Ctx,
+    ) -> Result<usize, E> {
+        self.start_offset += offset;
```
Comment (m4b) [inline src/buffer.rs:127]: hmm, why do you increment start offset by offset then subtract below? might be good to comment this since it isn't immediately clear

File: src/buffer.rs [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    /// use scroll::{DynamicBuffer, LE, Pwrite};
+    /// let mut buf = DynamicBuffer::new();
+    ///
+    /// buf.pwrite_with(2u32, 0, LE).unwrap();
+    /// assert_eq!(buf.get(), [2, 0, 0, 0]);
+    /// ```
+    pub fn get(&self) -> &[u8] {
+        &self.buffer[..=self.write_end]
+    }
+}
```
Comment (m4b) [inline src/buffer.rs:94]: I'm on fence about adding:
```rust
pub fn into_inner(self) -> Vec<u8> {
  self.buffer
}
```

on one hand it's nice and typical api of this sort; on other hand it forever forces the implementation to either use a Vec<u8> internally or create one on the fly if the user calls this method. It's probably fine?

File: src/buffer.rs [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+/// we cannot create a custom slice type that performs dynamic expansions under the hood.
+///
+/// As a result, we need to expose a separate writing type that can track offsets and calculate when to allocate.
+///
+/// However, if your `TryIntoCtx` impls don't use any special slice APIs and just use `Pwrite` and/or
+/// basic indexing, it's extremely easy to migrate! Just add the `DynamicBuffer` generic type.
+pub struct DynamicBuffer {
+    buffer: Vec<u8>,
+    alloc_increment: usize,
+    start_offset: usize,
```
Comment (m4b) [inline src/buffer.rs:63]: reading through this it isn't exactly clear why start_offset is for, and why the extra bookkeeping associated with it; i'm sure it is required, but might be nice to have some comments explaining it's use, etc.

File: src/pwrite.rs [hunk 7]
```diff
@@ -24,12 +24,11 @@ use crate::error;
 /// have a data that is not `&[u8]` — you will probably want to implement
 /// [TryIntoCtx](ctx/trait.TryIntoCtx.html) on your Rust types to be written.
 ///
-pub trait Pwrite<Ctx, E>: Index<usize> + IndexMut<RangeFrom<usize>> + MeasureWith<Ctx>
-where
-    Ctx: Copy,
-    E: From<error::Error>,
-{
-    fn pwrite<N: TryIntoCtx<Ctx, <Self as Index<RangeFrom<usize>>>::Output, Error = E>>(
+pub trait Pwrite<Ctx: Copy, E> {
```
Comment (m4b) [inline src/pwrite.rs:27]: Kind of talking to myself: So re my worries about dispatch ambiguity, i think it's not possilble; since implementors (e.g., blanket impl + DynamicBuffer) implement Pwrite, etc., and so disaptch is always `thing_that_implements_pwrite.pwrite` so there won't be any dispatch ambiguity, so yea that's a worry I don't need to have.

File: src/pwrite.rs [hunk 8]
```diff
@@ -24,12 +24,11 @@ use crate::error;
...
+    /// The type used by implementors of `TryIntoCtx` to write into your data store.
+    type Buffer: ?Sized;
```
Comment (m4b) [inline src/pwrite.rs:29]: so this part is peculiar to me; i have to think about it more; specifically, I'm wondering why this associated type is required; in e.g., your impl Buffer is Self, and pwrite requires basically `TryIntoCtx<Ctx, Self::Buffer, Error>`, which means `TryIntoCtx<Ctx, Self, Error>` so why wouldn't the associated type be dropped and we just pass Self there always?

I'm sure there's a good reason, but it would be nice if possible not to introduce the associated type, though if it is requried I think it's fine. I don't know of anyone putting explicit `Pwrite` bounds on their generic types (it shoudl only require `TryIntoCtx` bounds in order for `Pwrite` to work.

Quite an interesting solution to the problem you've done!

File: src/pwrite.rs [hunk 9]
```diff
@@ -39,29 +38,24 @@ where
...
         ctx: Ctx,
-    ) -> result::Result<usize, E> {
-        let len = self.measure_with(&ctx);
-        if offset >= len {
-            return Err(error::Error::BadOffset(offset).into());
-        }
-        let dst = &mut self[offset..];
-        n.try_into_ctx(dst, ctx)
-    }
+    ) -> result::Result<usize, E>;
```
Comment (m4b) [inline src/pwrite.rs:54]: so I don't like that this default impl is removed, but then I forgot there is a blanket impl of Pwrite below anyway, which is empty, because of this default impl; and this just gets moved there. The logic is a bit more separated, but it makes custom container types more flexible, which is quite cool.

File: src/pwrite.rs [hunk 10]
```diff
@@ -97,4 +92,19 @@ impl<
         R: ?Sized + Index<usize> + IndexMut<RangeFrom<usize>> + MeasureWith<Ctx>,
```
Comment (m4b) [inline src/pwrite.rs:92]: one thing that's puzzling me is how this blanket impl doesn't overlap the implementation of Pwrite for DynamicBuffer? (i.e., violates single implementation rule)

File: src/lib.rs [hunk 11]
```diff
@@ -229,6 +231,8 @@ mod lesser;
 mod pread;
 mod pwrite;
 
+#[cfg(feature = "std")]
+pub use crate::buffer::*;
```
Comment (m4b) [inline src/lib.rs:235]: what's funny in scroll prehistory is it had a module named `buffer` which did something similar (but it was just wrapping `&[u8]` basically, and wasn't dynamic), before I realized I could do everything generically with existing types.

File: src/lib.rs [hunk 12]
```diff
@@ -219,6 +219,8 @@ pub use scroll_derive::{IOread, IOwrite, Pread, Pwrite, SizeWith};
 #[cfg(feature = "std")]
 extern crate core;
 
+#[cfg(feature = "std")]
+mod buffer;
```
Comment (m4b) [inline src/lib.rs:223]: we may want to add a feature flag for dynamic buffers, not sure.

File: src/buffer.rs [hunk 13]
```diff
@@ -0,0 +1,294 @@
+use crate::ctx::TryIntoCtx;
```
Comment (m4b) [inline src/buffer.rs:1]: I think this entire module could actually be a third party crate; it is a self-defined type which implements a foreign trait, so ti won't violate orphan rules. (assuming we merge your Pwrite/Pread changes).

**Review by m4b — COMMENTED** [review]
This is pretty awesome! I'm quite surprised that tests pass, without any changes, honestly, although it is technically a breaking change (which is fine).

So, some thoughts:
1. I think the associated type Buffer is unavoidable? (which is fine if so, this is the technically breaking change)
2. Should Pread be updated (I don't think your DynamicBuffer works with pread? It might be nice from a symmetric viewpoint to get that working, but I don't think its' necessary)
3. There's also the argument to be made that this type could be in a third party crate (assuming the Pwrite changes go in)

The one thing I don't like is the asymmetry now between Pwrite/Pread; one of them has an associated type and the other doesn't. Is there a way to rectify this, you think?

Also what are you thoughts on DynamicBuffer being in a third party crate? (I think it's possible, I don't see an orphan rule violation anywhere). This would also punt on deciding on whether it should be under a feature flag (I generally don't like feature flags if can be avoided).

Lastly, w.r.t. the scroll_derive issues, I have to think about it, but I don't think we should generate extra implementations for DynamicBuffer in the default derive, since for my guess is majority of usecases won't require a DynamicBuffer, and it's just extra code bloat having them compile it.

Anyway, I don't see any _immediate_ issues preventing this from being merged; it's quite cool, and thanks for the PR; as I said, will need to be a little patient getting this in, don't have the most time right now, but I don't see any major blocking issues, however I would like to play with it a bit see if something comes up.

Anyway, feel free to add your thoughts here, etc. :) really great stuff :)

**Comment by m4b** [comment]
also it doesn't look like there's any CI anymore on scroll which is terrifying? (it used to run on jenkins, but it is probably dead?)

File: src/pwrite.rs (hunk 8 above)
Comment (nickbclifford) [reply src/pwrite.rs:29]: That's a good point, and I think you might be right. I added an associated type to make sure that the blanket implementations that already existed wouldn't change behavior, but if we change that to just a `impl [u8]` or something, then that shouldn't be a problem and we can get rid of it.

File: src/pwrite.rs (hunk 10 above)
Comment (nickbclifford) [reply src/pwrite.rs:92]: `DynamicBuffer` doesn't have a range index or implement `MeasureWith`.

**Comment by nickbclifford** [comment]
I'd be okay with moving `DynamicBuffer` into a separate crate, I believe all that would be necessary are the changes for Pwrite.

However, I do agree that the asymmetry with Pread is annoying, so I'll change that as well.

**Comment by nickbclifford** [comment]
Let me know what you think about the trait changes, and I'd be happy to move the buffer stuff out of this PR and change scope.

**Comment by m4b** [comment]
I have not forgotten about this! I will try to review tomorrow, thanks for your patience; always feel free to ping me if you aren't hearing from me, also :)

**Comment by nickbclifford** [comment]
@m4b Any thoughts about merging soon?

**Comment by m4b** [comment]
@nickbclifford ok, so here's what we should do:

1. if you can squash all your commits to single, and force push, that will make for a nice final review
2. I am inclined to merge this breaking change, without the `type Buffer` associated type; although this prevents some uses of const buffers, this may not ever matter, it would be an interesting case study to examine the assembly output to see if rustc is able to do anything with array vs slice

Unrelated to your patch, but you probably want a 0.11 crates release at some point:
1. i will likely remove the `usize` impl for `pread` and `pwrite` #56 
2. i will possibly look at making `pwrite` take `&` references, which is something i wanted for some time, though it may be a huge change, which is related to #51 and was the PR that never got merged here #47 

I would really want to revisit `pwrite` taking `&` before releasing 0.11, which at this stage would be effectively a beta for scroll 1.0. Before 1.0 i think pwrite taking references is good future proof design.

So in summary, if you can squash your changes, I think i'm ok to merge, and then i'd like to see at least 1. (but probably not 2., i'm not sure who would do it and i don't have time right now) before releasing 0.11

Does that sound reasonable? And once again, thanks for your patience here :)

**Comment by nickbclifford** [comment]
Yep, sounds all good to me!

**Review by m4b — APPROVED** [review]
LGTM! So thinking more about doing:

```rust
impl<Ctx: Copy, E: From<error::Error>, const LEN: usize> Pwrite<Ctx, E> for [u8; LEN] {
```
the problem here is that we want the buffer we write _into_ to be at most the size of the type; but we don't know the type until `pwrite` is called, e.g.:

```rust
fn pwrite_with<N: TryIntoCtx<Ctx, Self, Error = E>>(
```
which means we'd want something like:

```rust
fn pwrite_with<N: TryIntoCtx<Ctx, [u8; size_of::<N>()], Error = E>>(
```

which means the pwrite dispatches _from_ a buffer of some size N, _into_ a buffer that can contain the type.

If we just dispatch to the a slice, the value of const generics is effectively nil, since it is just a check on some constant `N` instead of `self.len()`.

What is interesting here is I don't think this pattern is expressible in scroll as defined, nor will an associated `type Buffer` help, since it's size is given in the impl block, but the size is effectively only known at `pwrite` definition time.

I think the only solution here would be to add another generic type to `Pwrite` which is a marker for the written to buffer? I'm not sure, have to think about it, but I've convinced myself that adding `type Buffer` adds no real benefit, other than allowing to dispatch _from_ const arrays to slices, which has minimal benefit.

Anyway, great PR, great discussion, and thanks for your patience!
